README.md

# SimpleDialog.js

This is just light weight dialog script inspired by jQueyUI's dialog widget.

jQueryUI is awesome library, but quite heavy, then I wrote this for simple usage.  
This doesn't have much feature like image gallery, dragging, or iframe content.  
If you want them, use jQuery UI, colorbox or something better.


## Usage

### $.fn.dialog( action:String, options:Object )

```javascript
var $dialog = $( /* some element */ );

$dialog.dialog("open", {
    width: 320,
    overlayColor: "#fff",
    overlayOpacity: 0.5
});

$dialog.on("dialogClose", function(){
    console.log("Dialog is closed");
});
```

### Actions

- "init" : Initialize dialog, doesn't open yet.
- "open" : Open the dialog. Before opening, initialize if not yet.
- "close" : Close the dialog
- "destroy" : Remove dialog node and overlay.

"init" and "open" accept options as argument.

### Options

- className:String ("simple-dialog") : Class name which is set to dialog element
- width:Integer (640) : Dialog width
- overlayColor:String (#000) : Overlay color
- overlayOpacity:Number (0.8) : Overlay opacity
- zIndex:Integer (1000) : Overlay z-index
- effect:String ("fade") : Effect method name
- openEasing:String ("swing") : Easing function name for opening dialog
- closeEasing:String ("swing") : Easing function name for closing dialog
- destroyOnClose:Boolean (false) : Remove dialog and overlay element when closed
- openDuration:Integer (300) : Duration time for opening dialog
- closeDuration:Integer (100) : Duration time for closing dialog
- closeButton:String (".button-close") : Selector expression for close button
- closeOnClickOverlay:Boolean (true) : Whether close the dialog when overlay is clicked or not

### Effects

SimpleDialog has only 3 effects.

- "fade" : Fade in and out
- "slide" : Similar to jQueryUI's "Clip" effect, buy more simple
- "none" : Imidiately show the dialog without animation

### Events

The element which is initialized as SimpleDialog triggered two kinds of events.

- "dialogOpen" : Fired when the dialog opened
- "dialogClose" : Fired when the dialog closed


## Author

mach3 @ <http://github.com/mach3>

- [Blog](http://blog.mach3.jp)
- [Website](http://www.mach3.jp)
- [Twitter](http://twitter.com/mach3ss)

doc/CHANGELOG.md

# CHANGE LOG

## 0.9.1 (2013/08/06)

- Add options (destroyOnClose, zIndex)
- Fix Bugs
- Fix Mistakes in README


src/simpledialog.effects.js
(function($){

	/**
	 * Effect functions for $.SimpleDialog
	 * Add new effect by extending $.SimpleDialog.prototype.effects
	 */
	var effects = $.SimpleDialog.prototype.effects;

	effects.slide = function(open){
		var options = this.getEffectOptions(open);
		if(open){
			options.always = function(){
				$(this).css("overflow", "visible");
			};
			this.dialog.stop().css({
				visibility: "visible",
				overflow: "hidden",
				top: this.offset.top + this.offset.height / 2,
				height: 0
			})
			.animate({
				top: this.offset.top,
				height: this.offset.height
			}, options);
		} else {
			options.always = function(){
				$(this).css({
					visibility: "hidden",
					height: "auto"
				});
			};
			this.dialog.stop().animate({
				top: this.offset.top + this.offset.height / 2,
				height: 0
			}, options);
		}
	};

}(jQuery));

src/simpledialog.js
(function($, win, doc){

	/**
	 * SimpleDialog class
	 * ------------------
	 */
	$.SimpleDialog = function($node){
		this.initialize($node);
	};
	(function(){

		/**
		 * Attributes:
		 * - node:jQueryObject 
		 * - dialog:jQueryObject
		 * - overlay:jQueryObject
		 * - offset:Object
		 */
		this.node = null;
		this.dialog = null;
		this.overlay = null;
		this.offset = null;

		/**
		 * Actions available in $.fn.dialog()
		 */
		this.actions = [
			"init",
			"open", 
			"close",
			"destroy"
		];

		/**
		 * Options:
		 * - className:String = CSS class name set for dialog
		 * - width:Integer = Dialog width
		 * - overlayColor:String = Overlay background color
		 * - overlayOpacity:Number = Overlay opacity
		 * - effect:String = Effect method name for opening/closing dialog
		 * - openEasing:String = Easing function name for opening dialog
		 * - openDuration:Integer = Duration time for opening dialog
		 * - closeEasing:String = Easing function name for closing dialog
		 * - closeDuration:Integer = Duration time for closing dialog
		 * - closeButton:String = Selector expression for close button
		 * - closeOnClickOverlay:Boolean = Whether close dialog when overlay is clicked or not
		 */
		this.options = {
			className: "simple-dialog",
			width: 640,
			overlayColor: "#000",
			overlayOpacity: 0.8,
			zIndex: 1000,

			effect: "fade",
			openEasing: "swing",
			openDuration: 300,
			closeEasing: "swing",
			closeDuration: 100,
			destroyOnClose: false,

			closeButton: ".button-close",
			closeOnClickOverlay: true
		};

		/**
		 * Effects for opening/closing dialog
		 * These are assigned $.SimpleDialog instance as "this"
		 * An argument passed is boolean for opening or closeing dialog
		 */
		this.effects = {
			"fade": function(open){
				var options = this.getEffectOptions(open);
				if(open){
					this.dialog.css({
						visibility: "visible",
						opacity: 0
					})
					.animate({ opacity: 1 }, options);
				} else {
					this.dialog.animate({ opacity: 0 }, options);
				}
			},
			"none": function(open){
				var visible = open ? "visible" : "hidden";
				this.dialog.css("visibility", visible);
			}
		};

		/**
		 * Initialize instance
		 * @constructor
		 * @param jQueryObject node
		 */
		this.initialize = function(node){
			this.node = node;
			this.options = $.extend({}, this.options);
		};

		/**
		 * Initialize dialog and overlay
		 * @param Object options
		 */
		this.init = function(options){
			this.config(options);
			this.getOverlay();
			this.getDialog();
		};

		/**
		 * Open the dialog
		 * @param Object options
		 */
		this.open = function(options){
			this.init(options);
			this.overlay.show();
			this.dialog.css(this.getOffset());
			this.effects[this.get("effect")].call(this, true);
		};

		/**
		 * Close the dialog
		 */
		this.close = function(){
			var my = this;
			this.effects[this.get("effect")].call(this, false);
			if(this.get("destroyOnClose")){
				this.node.on("dialogClose", function(){
					my.destroy();
				});
			}
			this.overlay.hide();
		};

		/**
		 * Remove overlay and dialog node
		 */
		this.destroy = function(){
			this.overlay.remove();
			this.dialog.remove();
		};

		/**
		 * Configure options
		 * @param Object options
		 */
		this.config = function(options){
			$.extend(this.options, options);
		};

		/**
		 * Get value from options by key
		 * @param String key
		 */
		this.get = function(key){
			return this.options[key];
		};

		/**
		 * Get or create overlay node, set it as instance member
		 * @return jQueryObject
		 */
		this.getOverlay = function(){
			if(! this.overlay){
				this.overlay = $("<div>")
				.css({
					display: "none",
					position: "absolute",
					left: 0,
					top: 0,
					width: "100%",
					height: "100%",
					backgroundColor: this.get("overlayColor"),
					opacity: this.get("overlayOpacity"),
					zIndex: this.get("zIndex")
				})
				.prependTo("body");
				if(this.get("closeOnClickOverlay")){
					this.overlay.on("click", $.proxy(this.close, this));
				}
			}
			return this.overlay;
		};

		/**
		 * Get or create dialog node, set it as instance member
		 * @return jQueryObject
		 */
		this.getDialog = function(){
			if(! this.dialog){
				this.dialog = $("<div>")
				.css({
					position: "absolute",
					visibility: "hidden",
					width: this.get("width"),
					zIndex: this.get("zIndex") + 1
				})
				.addClass(this.get("className"))
				.append(this.node)
				.insertAfter(this.overlay)
				.on("click", this.get("closeButton"), $.proxy(this.close, this));
			}
			return this.dialog;
		};

		/**
		 * Calculate width, height and position, set them as instance member
		 * @retnrn Object
		 */
		this.getOffset = function(){
			var client = {
				scroll: $(window).scrollTop(),
				width: win.innerWidth || doc.documentElement.clientWidth,
				height: win.innerHeight || doc.documentElement.clientHeight
			};
			this.offset = {
				width: this.dialog.width(),
				height: this.dialog.height()
			};
			this.offset.left = (client.width - this.offset.width) / 2;
			this.offset.top = ((client.height - this.offset.height) / 2) + client.scroll;
			return this.offset;
		};

		/**
		 * Get options for $.fn.animate()
		 * Duration, easing and callback set in options
		 * `open` is opening or closing dialog
		 * @param Boolean open
		 */
		this.getEffectOptions = function(open){
			var prefix, my;
			prefix = open ? "open" : "close";
			my = this;
			return {
				duration: this.get(prefix + "Duration"),
				easing: this.get(prefix + "Easing"),
				complete: function(){
					if(! open){
						$(this).css("visibility", "hidden");
					}
					my.node.trigger("dialog" + my._capitalize(prefix));
				}
			};
		};

		/**
		 * Capitalize string
		 * @param String str
		 * @return String
		 */
		this._capitalize = function(str){
			return str.replace(/^[a-z]/, function(s){
				return s.toUpperCase();
			})
		};

	}).call($.SimpleDialog.prototype);


	/**
	 * Create dialog or operate dialog by passing named action
	 * @param String action
	 * @param Object options
	 * @return jQueryObject
	 */
	$.fn.dialog = function(action, options){
		var dialog;
		if(! this.data("dialog")){
			this.data("dialog", new $.SimpleDialog(this));
		}
		dialog = this.data("dialog");
		if($.inArray(action, dialog.actions) >= 0){
			dialog[action](options);
		}
		return this;
	};

}(jQuery, window, document));